### PR TITLE
Diet Rework

### PR DESCRIPTION
**Pull request by pstalcup**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -264,6 +264,7 @@ export function main(argString = ""): void {
   }
 
   const args = argString.split(" ");
+  let simulateDiet = false;
```
Comment (horrible-little-slime) [inline src/index.ts:267]: put this in globalOptions

Comment (horrible-little-slime) [reply src/index.ts:267]: I guess it's probably temporary for diagnostic stuff, so this is fine.

But if you want it to be a permanent feature, or if you are going to transform this architecture into `nodiet`, make it part of globalOptions.

Comment (phulin) [reply src/index.ts:267]: why not just put the runDiet call inside the if statement? I don't think any other args will interact with this one

File: src/diet.ts [hunk 2]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
-    new MenuItem($item`tin cup of mulligan stew`),
-    new MenuItem($item`glass of raw eggs`, { maximum: availableAmount($item`glass of raw eggs`) }),
-    new MenuItem($item`Tea, Earl Grey, Hot`),
-    new MenuItem($item`meteoreo`),
-    new MenuItem($item`ice rice`),
-    new MenuItem($item`frozen banquet`),
-    new MenuItem($item`fishy fish lasagna`),
-    new MenuItem($item`gnat lasagna`),
-    new MenuItem($item`long pork lasagna`),
+    new MenuItem<string>($item`Dreadsylvanian spooky pocket`),
```
Comment (phulin) [inline src/diet.ts:267]: you don't need the explicit type parameters here - TS can infer from the return value.

File: src/diet.ts [hunk 3]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
+    (get("beGregariousFightsLeft") > 0 &&
+      get("beGregariousMonster") === $monster`Knob Goblin Embezzler`);
+
+  return [
+    new MenuItem<string>($item`Extrovermectin™`, {
+      additionalValue: expectedGregs() * embezzlerDifferential,
+      maximum: alreadyGregarious ? 0 : 1,
+      data: "first gregarious",
+    }),
+    /*new MenuItem<string>($item`Extrovermectin™`, {
```
Comment (phulin) [inline src/diet.ts:345]: no commented code please :)

File: src/diet.ts [hunk 4]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
-    if (haveEffect(effect) < effectDuration) {
-      // Adjust value of horseradish if we have mayo, since we're using Mayozapine instead of Mayoflex.
-      const firstValue =
-        value + meatDrop * Math.min(embezzlers, effectDuration) * 750 - (useMayo ? MPA : 0);
-      menu.push(new MenuItem(item, { maximum: 1, additionalValue: firstValue }));
-    }
-  }
+function copiers() {
+  const embezzlerDifferential = 25000 - 7500;
+  // const normalGregs = have($item`miniature crystal ball`) ? 4 : 3;
```
Comment (phulin) [inline src/diet.ts:333]: no commented code please :)

File: src/diet.ts [hunk 5]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
-  // We don't have a property to check if nothing has been eaten, so use this hack.
-  // This will fail in the rare case where someone has eaten non-PVPable food
-  // and then cleansed back down to 0.
-  if (
-    have($item`spaghetti breakfast`) &&
-    myFullness() === 0 &&
-    get("_timeSpinnerFoodAvailable") === "" &&
-    !get("_spaghettiBreakfastEaten")
+function countCopies(diet: Diet<string>) {
+  // const normalGregs = have($item`miniature crystal ball`) ? 4 : 3;
```
Comment (phulin) [inline src/diet.ts:360]: same

File: src/diet.ts [hunk 6]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
-  const spleenCleaner = argmax(spleenCleaners.map((item) => [item, mallPrice(item)]));
 
-  const embezzlers = embezzlerCount();
-  const helpers = [Mayo.flex, $item`Special Seasoning`, saladFork, frostyMug];
-  const menu = [
+/**
+ * Generate a basic menu of high-yield items to consider
+ * @returns basic menu
+ */
+function menu(): MenuItem<string>[] {
```
Comment (phulin) [inline src/diet.ts:256]: Can you make the type parameter record what the possible values are? I.e. define `Note` to be a union of whatever the tiers are, "first gregarious", etc.

File: src/diet.ts [hunk 7]
```diff
@@ -342,130 +240,308 @@ function pillCheck(): void {
...
+    const valuestr = `value: ${Math.floor(
+      dietEntry.expectedValue(MPA, diet)
+    )}${addvalstr} price: ${Math.floor(dietEntry.expectedPrice())}`;
+    print(`${dietEntry.quantity}${maxstr} ${target}${datastr}${helpersstr} ${valuestr}`);
+  }
+  const totalValue = diet.expectedValue(MPA);
+  const totalCost = diet.expectedPrice();
+  const netValue = totalValue - totalCost;
+  print(
+    `Assuming MPA of ${MPA}. Total Cost: ${totalCost}, Total Value ${totalValue}, Net Value ${netValue}`
```
Comment (phulin) [inline src/diet.ts:538]:
```suggestion
    `Assuming MPA of ${MPA}. Total Cost ${totalCost}, Total Value ${totalValue}, Net Value ${netValue}`
```

File: src/diet.ts [hunk 8]
```diff
@@ -478,85 +554,24 @@ function itemPriority(menuItems: MenuItem[]) {
...
-
-  // Sort descending by item priority.
-  diet.sort(([x], [y]) => -(itemPriority(x) - itemPriority(y)));
-
+export function consumeDiet(diet: Diet<string>): void {
+  diet = diet.copy();
   print();
   print("===== PLANNED DIET =====");
   printDiet(diet);
   print();
```
Comment (phulin) [inline src/diet.ts:562]: can we skip the printing if the diet's empty?

**Review by phulin — CHANGES_REQUESTED** [review]

File: src/diet.ts [hunk 9]
```diff
@@ -342,130 +241,317 @@ function pillCheck(): void {
...
+    ...[...stomachLiverCleaners.keys()].map((item) => new MenuItem<Note>(item)),
+  ];
+}
+
+function gregariousCopies(): number {
+  return 3 + (have($item`miniature crystal ball`) ? 1 : 0);
+}
+
+function copiers(): MenuItem<Note>[] {
+  const embezzlerDifferential = 25000 - 7500;
```
Comment (horrible-little-slime) [inline src/diet.ts:338]:
```suggestion
  const embezzlerDifferential = 25000 - MPA;
```

Comment (horrible-little-slime) [reply src/diet.ts:338]: Wish we had a better estimate than 25k but that's probably fine

File: src/diet.ts [hunk 10]
```diff
@@ -342,130 +241,317 @@ function pillCheck(): void {
...
+    diet: () => fullDietPlanner(balanceMenu(menu(), fullDietPlanner)),
+    shotglass: () =>
+      shotglassDietPlanner(
+        balanceMenu(
+          menu().filter((menuItem) => itemType(menuItem.item) === "booze" && menuItem.size === 1),
+          shotglassDietPlanner
+        )
+      ),
+    pantsgiving: () =>
+      pantsgivingDietPlanner(
```
Comment (horrible-little-slime) [inline src/diet.ts:514]: It feels like overkill to use the diet code for this, but I guess it probably doesn't cost us that much

File: src/diet.ts [hunk 11]
```diff
@@ -612,37 +635,56 @@ export function runDiet(): void {
             }
           }
           consumeSafe(countToConsume, menuItem.item);
-        } else if (menuItem.item === Mayo.flex) {
-          if (menuItems[menuItems.length - 1].item === $item`jumping horseradish`) {
-            MayoClinic.setMayoMinder(Mayo.zapine, countToConsume);
-          } else {
-            MayoClinic.setMayoMinder(Mayo.flex, countToConsume);
-          }
+        } else if ([Mayo.flex, Mayo.zapine].includes(menuItem.item)) {
```
Comment (horrible-little-slime) [inline src/diet.ts:638]: I know that we don't currently support mayodiol, but you might do like `Array.from(Object.entries(Mayo)).includes(menuItem.item)` if you feel like being terrible

File: src/diet.ts [hunk 12]
```diff
@@ -612,37 +635,56 @@ export function runDiet(): void {
...
+    if (!get("_mimeArmyShotglassUsed")) {
+      consumeDiet(dietBuilder.shotglass());
+    }
+
+    if (
+      get("barrelShrineUnlocked") &&
+      !get("_barrelPrayer") &&
+      $classes`Turtle Tamer, Accordion Thief`.includes(myClass())
+    ) {
+      cliExecute("barrelprayer buff");
```
Comment (horrible-little-slime) [inline src/diet.ts:685]: dumb question: is it worth checking to see if we actually plan on using enough fullness in the associated organ before pulling the trigger on this?

File: src/lib.ts [hunk 13]
```diff
@@ -429,6 +433,10 @@ export function printHelpMenu(): void {
     +--------------+---------------------------------------------------------------------------------------------------+
     | &lt;somenumber&gt; | garbo will terminate after the specified number of turns, e.g. \`garbo 200\` will terminate after   |
     |              |  200 turns are spent. Negative inputs will cause garbo to terminate when the specified number of turns remain.       |
+    +------------------------------------------------------------------------------------------------------------------+
+    |   simdiet    | garbo will print out what it computes as an optimal diet and then exit                            |
+    +------------------------------------------------------------------------------------------------------------------+
+    |    nodiet    | *EXPERIMENTAL* garbo will not eat or drink anything as a part of its run (including pantsgiving)  |
```
Comment (horrible-little-slime) [inline src/lib.ts:439]: add these to README, if you dare

Comment (horrible-little-slime) [reply src/lib.ts:439]: We can wait to push before doing this; you can declare this out of scope of this PR.

File: src/post.ts [hunk 14]
```diff
@@ -41,13 +38,8 @@ function coldMedicineCabinet(): void {
 }
 
 function horseradish(): void {
```
Comment (horrible-little-slime) [inline src/post.ts:40]: Should probably rename this function if it's not always a horseradish!

File: src/potions.ts [hunk 15]
```diff
@@ -337,7 +339,7 @@ export function potionSetup(embezzlersOnly: boolean): void {
   const testPotions = farmingPotions.filter(
     (potion) => potion.gross(embezzlers) / potion.price(true) > 0.5
   );
-  testPotions.sort((x, y) => -(x.net(embezzlers) - y.net(embezzlers)));
+  testPotions.sort((a, b) => b.net(embezzlers) - a.net(embezzlers));
```
Comment (horrible-little-slime) [inline src/potions.ts:342]: Thank you for that stylistic change

File: src/potions.ts [hunk 16]
```diff
@@ -274,7 +276,7 @@ function useAsValuable(potion: Potion, embezzlers: number, embezzlersOnly: boole
   const value = potion.value(embezzlers);
   const price = potion.price(false);
   const amountsAcquired = value.map((value) =>
-    (!embezzlersOnly || value.name === "embezzlers") && value.value - price > 0
+    (!embezzlersOnly || value.name === "embezzler") && value.value - price > 0
       ? acquire(value.quantity, potion.potion, value.value, false)
```
Comment (horrible-little-slime) [inline src/potions.ts:280]: I am always concerned about things like `value.value` being written down--I think we can declare cleanup out of scope, but I would like to try to make this a little more legible for future generations in the future.

**Review by horrible-little-slime — APPROVED** [review]
I'd like to get the hot new libram off the presses before pushing, because that was a shitton of mallsearches, but this is working fine for me right now, and boy howdy is it impressive work.

File: src/potions.ts (hunk 16 above)
Comment (phulin) [reply src/potions.ts:280]: agreed

**Pull request merged**